### PR TITLE
Authenticate a registered graph from explorer

### PR DESCRIPTION
**Pull request by mayakoneval**
**Recommend reviewing with whitespace off**

**Context**:

[JIRA
](https://apollographql.atlassian.net/browse/NEBULA-1090)

[Slack conversation 1
](https://apollograph.slack.com/archives/C032ZPPE5E1/p1652923931401779)

[Slack conversation 2
](https://apollograph.slack.com/archives/C032ZPPE5E1/p1652996965376389)

In the same way the code in the `embeddable-explorer` repo is parent code for the parent rendering the embed to run, dev tools runs the same (or very similar code) to be the parent. We can't use the code in `embeddable-explorer` because dev tools sends requests differently that a parent does. Dev tools send requests through the AC instance, a parent just `fetch`s.

**This PR**

1. Add a query param to hide the explorer header in dev tools
2. Makes some general clean up to dev tools including formatting all of the files & moving constants into their own folder
3. If we can't introspect your schema, show a modal asking you to input a graph ref so we can use the registered schema from Studio. 
4. When you have chosen a graph, expose a button that lets you switch the graph if you want
5. Save the last graph ref you used in local storage, so that if we can't introspect your schema again, we just use the last graph you inputted
6. If you have authenticated graph refs in the past and you re-input them, don't ask you to authenticate again

https://user-images.githubusercontent.com/14367451/169465033-37a516e1-cf1c-498e-87d8-a1417caf536c.mov

### CODE REVIEW

**Comment by mayakoneval** [comment]
@Jephuff any interest in reviewing this to ramp up lol?

**Comment by mayakoneval** [comment]
TODO (Maya) show a loading screen behind the Explorer so it looks like its loading when the iframe hasn't loaded yet

File: src/application/components/Explorer/Explorer.tsx [hunk 1]
```diff
@@ -174,64 +186,65 @@ export const Explorer = ({ navigationProps, embeddedExplorerProps }: {
       });
 
       observer.subscribe((response: QueryResult) => {
-        if(response.networkStatus === NetworkStatus.error) {
-          embeddedExplorerIFrame.contentWindow?.postMessage({
-            name: SCHEMA_ERROR,
-            schema: response.data,
```
Comment (mayakoneval) [inline src/application/components/Explorer/Explorer.tsx:180]: schema isn't necessary to send here - there is no schema - we don't expect this on the studio-ui side

File: src/application/components/Explorer/postMessageHelpers.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      graphRef: string;
+    }
+  | {
+      name: "PartialAuthenticationTokenResponse";
+      partialToken: string;
+    }
+  | {
+      name: "DevTools_AskForGraphRefAndAuthenticate";
+    };
+
```
Comment (mayakoneval) [inline src/application/components/Explorer/postMessageHelpers.ts:82]: added typings for incoming and outgoing messages in this file - before we were just winging it

File: src/application/components/Explorer/Telescope.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      <path
+        fill="#444"
+        d="M20.849 209.683C7.639 188.617 0 163.7 0 137 0 61.337 61.337 0 137 0s137 61.337 137 137c0 26.7-7.638 51.617-20.849 72.683C218.266 191.978 178.798 182 137 182s-81.266 9.978-116.151 27.683z"
+        opacity=".02"
+      />
+      <path fill="#E5358D" d="M111 140.5l5.5 12.5 6.1-2.5-5.3-12.7z" />
+      <path fill="#A7A7A7" d="M158 116.5l8.5 20L151 143l-8.5-20z" />
+      <path
+        fill="#44444D"
+        d="M150.955 144.734a2.038 2.038 0 0 1-1.877-1.223l-3.837-8.913a1.991 1.991 0 0 1 1.083-2.625c1.041-.432 2.232.045 2.67 1.064l3.041 7.068 8.685-3.61c1.04-.433 2.233.045 2.67 1.063a1.991 1.991 0 0 1-1.082 2.625l-10.56 4.393c-.26.107-.53.158-.793.158"
```
Comment (mayakoneval) [inline src/application/components/Explorer/Telescope.tsx:24]: copy pasted svg from studio

File: src/application/components/Explorer/postMessageAuthHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import { IncomingMessageEvent, postMessageToEmbed } from "./postMessageHelpers";
+
+export const SET_PARTIAL_AUTHENTICATION_TOKEN_FOR_PARENT =
+  "SetPartialAuthenticationTokenForParent";
+export const EXPLORER_LISTENING_FOR_PARTIAL_TOKEN =
+  "ExplorerListeningForPartialToken";
+export const PARTIAL_AUTHENTICATION_TOKEN_RESPONSE =
+  "PartialAuthenticationTokenResponse";
+
```
Comment (mayakoneval) [inline src/application/components/Explorer/postMessageAuthHelpers.ts:9]: most of this is copy pasted from studio: https://github.com/mdg-private/studio-ui/blob/main/src/app/embedHelpers/useListenForAuthResponse.tsx

File: src/application/components/Explorer/postMessageHelpers.ts [hunk 5]
```diff
@@ -129,20 +130,14 @@ type SetPartialAuthenticationTokenForParent = MessageEvent<{
   graphRef: string;
 }>;
 
-type DevToolsGraphRefAuthenticated = MessageEvent<{
```
Comment (mayakoneval) [inline src/application/components/Explorer/postMessageHelpers.ts]: forgot to delete this in the previous commit - from a previous iteration

File: src/application/components/Explorer/LoadingSpinner.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
+/** @jsx jsx */
```
Comment (mayakoneval) [inline src/application/components/Explorer/LoadingSpinner.tsx:1]: this file pretty much exactly copy pasted from studio-ui

File: src/application/components/Explorer/Explorer.tsx [hunk 7]
```diff
@@ -104,42 +116,88 @@ function executeOperation({
...
-    setEmbeddedExplorerIFrame: (iframe: HTMLIFrameElement) => void,
-  }
+    embeddedExplorerIFrame: HTMLIFrameElement | null;
+    setEmbeddedExplorerIFrame: (iframe: HTMLIFrameElement) => void;
+  };
 }): jsx.JSX.Element => {
-  const [schema, setSchema] = useState<IntrospectionQuery | null>(null)
+  const [graphRef, setGraphRef] = useState<string>();
+  const [showGraphRefModal, setShowGraphRefModal] = useState<
+    false | "triggeredByIntrospection" | "triggeredManually"
```
Comment (cheapsteak) [inline src/application/components/Explorer/Explorer.tsx:184]: minor nit, could we rename this to `triggeredByIntrospectionFailure` 🙏 was initially confused about why we'd want to show the modal when we receive a successful introspection

File: src/application/components/Explorer/GraphRefModal.tsx [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+import { Telescope } from "./Telescope";
+
+export const GraphRefModal = ({
+  onClose,
+  graphRef,
+  setGraphRef,
+  setNewGraphRefLoading,
+  newGraphRefLoading,
+  embeddedExplorerIFrame,
+  wasTriggeredByIntrospection,
```
Comment (cheapsteak) [inline src/application/components/Explorer/GraphRefModal.tsx:21]: ditto here 🙏

**Review by cheapsteak — APPROVED** [review]
looks great!

**Comment by brainkim** [comment]
Not blocking the review, just putting a pin in this so I can take a look at it later. Thanks!

**Pull request merged**